docs/developer-docs/security/_attachments/lock.rs

```diff
@@ -0,0 +1,40 @@
+use std::cell::RefCell;
+use std::cmp::Ord;
+use std::collections::BTreeSet;
+use std::rc::Rc;
+
+pub struct State<T: Ord> {
+    pending_requests: BTreeSet<T>,
+}
+
+impl<T: Ord> Default for State<T> {
+    fn default() -> Self {
+        Self {
+            pending_requests: BTreeSet::new(),
+        }
+    }
+}
+
+pub struct CallerGuard<T: Ord> {
+    state: Rc<RefCell<State<T>>>,
```

Comment:
Minor point: Is the Rc needed here? Maybe add a comment to explain why it is used and would be useful in this case?

Reply:
```suggestion
    /// We use `Rc` and `RefCell` here as we need both multiple owners (every lock is an owner) and mutable borrows for the state (updating the state after releasing the lock).
    state: Rc<RefCell<State<T>>>,
```